libs/core/langchain_core/runnables/base.py

```diff
@@ -659,16 +664,187 @@ async def astream_log(
+        | on_tool_end          | some_tool        |                                 |                                               | {"x": 1, "y": "2"}                              |
+        | on_retriever_start   | [retriever name] |                                 | {"query": "hello"}                            |                                                 |
+        | on_retriever_chunk   | [retriever name] | {documents: [...]}              |                                               |                                                 |
+        | on_retriever_end     | [retriever name] |                                 | {"query": "hello"}                            | {documents: [...]}                              |
+        | on_prompt_start      | [template_name]  |                                 | {"question": "hello"}                         |                                                 |
+        | on_prompt_end        | [template_name]  |                                 | {"question": "hello"}                         | ChatPromptValue(messages: [SystemMessage, ...]) |
+
+        Here are declarations associated with the events shown above:
+
+        `format_docs`:
+
+        ```python
+        def format_docs(docs: List[Document]) -> str:
+            '''Format the docs.'''
+            return ", ".join([doc.page_content for doc in docs])
+
+        format_docs = RunnableLambda(format_docs)
+        ```
+
+        `some_tool`:
+
+        ```python
+        @tool
+        def some_tool(x: int, y: str) -> dict:
+            '''Some_tool.'''
+            return {"x": x, "y": y}
+        ```
+
+        `prompt`:
+
+        ```python
+        template = ChatPromptTemplate.from_messages(
+            [("system", "You are Cat Agent 007"), ("human", "{question}")]
+        ).with_config({"run_name": "my_template", "tags": ["my_template"]})
+        ```
+
+        Example:
+
+        .. code-block:: python
+
+            from langchain_core.runnables import RunnableLambda
+
+            async def reverse(s: str) -> str:
+                return s[::-1]
+
+            chain = RunnableLambda(func=reverse)
+
+            events = [event async for event in chain.astream_events("hello")]
+
+            # will produce the following events (run_id has been omitted for brevity):
+            [
+                {
+                    "data": {"input": "hello"},
+                    "event": "on_chain_start",
+                    "metadata": {},
+                    "name": "reverse",
+                    "tags": [],
+                },
+                {
+                    "data": {"chunk": "olleh"},
+                    "event": "on_chain_stream",
+                    "metadata": {},
+                    "name": "reverse",
+                    "tags": [],
+                },
+                {
+                    "data": {"output": "olleh"},
```

Comment:
what does this look like for runs where the input is sent to tracer only at the end?

Reply:
Gets an additional `input` key. Data is always a dict with this format, so we can output independent things when relevant


```
"data": {"output": "olleh", "input": "hello"}
```